### PR TITLE
Advocate using locals to render views

### PR DESCRIPTION
**Pull request by mcmire**
It has [long] been [argued] that using instance variables to pass values
to Rails views isn't the greatest of ideas: Rails twists instance
variables to do something they aren't supposed to do, and if you think
about it, it's unclear how views even have access to variables that live
in a completely different scope. (Hint: they're copied from the
controller to the view at runtime when the view is instantiated.)

In most other languages, views (or templates, as they're sometimes
called) are essentially functions and, as such, when you're rendering a
template and want to give that template access to certain values, you
just pass those values along as arguments.

Is there a way to do this in Rails? Yes, by calling `render` explicitly
and using the `locals` option. So your controller would go from this:

``` ruby
class PostsController < ApplicationController
  def index
    @posts = Post.all
  end
end
```

to this:

``` ruby
class PostsController < ApplicationController
  def index
    posts = Post.all
    render locals: { posts: posts }
  end
end
```

The upside to this approach is that now, views and partials act the same
in terms of how they reference variables, and you can convert a view to
a partial easily.

The downside is that you can no longer easily assert that a view was
rendered with a particular variable, but that's okay, because we don't
write controller specs anyway.

[long]: https://github.com/hashrocket/decent_exposure
[argued]: http://thepugautomatic.com/2013/05/locals/

### CODE REVIEW

File: style/rails/sample.rb [hunk 1]
```diff
@@ -1,7 +1,15 @@
-class SomeClass
+class SomeModel < ActiveRecord::Base
+  belongs_to :tree
```
Comment (gylaz) [inline style/rails/sample.rb]: Is this unrelated?

**Comment by gylaz** [comment]
:+1: :100:

File: style/rails/sample.rb (hunk 1 above)
Comment (mcmire) [inline style/rails/sample.rb]: Yeah, it is. I can put it in another commit.

**Comment by derekprior** [comment]
I like this in theory but I feel like it's also a bit too far afield from Rails norms. Doing this an established client project would surely bring some conflict and a developer taking over one of our greenfield projects would likely be confused by this practice. 

I don't think this solves a big enough problem to justify so drastically veering from Rails norms.

We should probably try this on a few different types of projects if we want to consider making it part of guides. I haven't seen this in our code yet.

**Comment by tommydangerous** [comment]
This is interesting, I see it done in the `administrate` gem.

**Comment by mcmire** [comment]
> We should probably try this on a few different types of projects if we want to consider making it part of guides.

Sure. There is a [research card](https://trello.com/c/49MHwpol) open on this where people can try it out.

**Comment by jferris** [comment]
I'm going to close this while we're running the experiment. We can re-open if the experiment is successful.

**Comment by mcmire** [comment]
Gotcha, just wanted to get more eyes on it.

**Comment by Drenmi** [comment]
> The downside is that you can no longer easily assert that a view was
rendered with a particular variable, but that's okay, because we don't
write controller specs anyway.

For what it's worth, it looks like [`assigns` will be deprecated in Rails 5](https://github.com/rails/rails/issues/18950).

**Comment by jferris** [comment]
I just noticed this in an earlier comment:

> We don't write controller specs anyway.

There isn't anything in our guides that says not to write controller specs, and I use them reasonably often. Unless we come to a consensus that controller specs are not worth supporting, I wouldn't make other changes that assume that nobody is writing them.

**Comment by gylaz** [comment]
> We don't write controller specs anyway.

Is this a reason to use `render locals...`? Is there a way to still access (in tests) these variables? I feel like there must be, especially if `assigns` is deprecated in Rails 5.

**Comment by jferris** [comment]
> Is this a reason to use `render locals...`? Is there a way to still access (in tests) these variables? I feel like there must be, especially if `assigns` is deprecated in Rails 5.

I don't think I've recently written a test that makes an assertion on `assigns`. I test inputs and outputs to controllers in controller specs. I usually test side effects like sending emails. When testing responses, I generally just check whether a certain template was rendered or an action redirected.

**Comment by dideler** [comment]
Curious to hear if the experiment is successful. This is a change I've been personally wanting to do but have been cautious for the same reasons mentioned in https://github.com/thoughtbot/guides/pull/398#issuecomment-174331194.

**Pull request closed without merging**